File: heap/0313_super_ugly_number/0313_super_ugly_number.go
package main 

import (
	"fmt"
	"math"
)

func nthSuperUglyNumber(n int, primes []int) int {
	ugly := make([]int, n)
	idx := make([]int, len(primes))
	val := make([]int, len(primes))
	for i := 0; i < len(primes); i++ {
		val[i] = 1
	}

	next := 1
	for i := 0; i < n ; i++ {
		ugly[i] = next
		next = math.MaxInt32
		for j := 0; j <len(primes); j++ {
			if val[j] == ugly[i] {
				val[j] = ugly[idx[j]] * primes[j]
				idx[j]++
			}
			next = min(next, val[j])
		}
	}
	return ugly[n-1]
}

func min(x,y int) int {
	if x < y {
		return x
	} 
	return y
}

func main() {
	n := 12
	primes := []int{2,7, 13, 19}
	res := nthSuperUglyNumber(n, primes)
	fmt.Println(res)
}

File: heap/0313_super_ugly_number/0313_super_ugly_number.py
import heapq


class Solution(object):
    def nthSuperUglyNumber(self, n, primes):
        heap, uglies, idx, ugly_by_last_prime = [], [0] * n, [0] * len(primes), [0] * n
        uglies[0] = 1

        for k, p in enumerate(primes):
            heapq.heappush(heap, (p, k))

        for i in xrange(1, n):
            uglies[i], k = heapq.heappop(heap)
            ugly_by_last_prime[i] = k
            idx[k] += 1
            while ugly_by_last_prime[idx[k]] > k:
                idx[k] += 1
            heapq.heappush(heap, (primes[k] * uglies[idx[k]],k))

        return uglies[-1]


n = 12
primes = [2,7,13,19]
res = Solution().nthSuperUglyNumber(n, primes)
print(res)